### PR TITLE
Site Settings: Add Disconnect Confirmation Page

### PR DESCRIPTION
**Pull request by ockham**
Based on and supersedes @AnnaMag's #18039.

Add a route ` /settings/disconnect-site/confirm/:site` that displays the following:

image

Note that nothing currently points to that route; however, it will be the final step of the Disconnect JP Survey which I'll be adding with #18749.

To test:
* Verify that 'Stay Connected' takes you to `settings/manage-connection/:site`.
* Verify that 'Disconnect' actually disconnects your JP site, and redirects you to 'Stats'
* Try the route for a non-JP site; you should be redirected to `/settings/general`.

### CODE REVIEW

File: client/my-sites/site-settings/index.js [hunk 1]
```diff
@@ -63,6 +63,14 @@ export default function() {
 			settingsController.setScroll,
 			controller.disconnectSite
 		);
+
+		page(
+			'/settings/disconnect-site/:site_id/confirm',
```
Comment (tyxla) [inline client/my-sites/site-settings/index.js]: We've been trying to keep the site fragment at the end of routes for various reasons (consistency, and for helping some utilities do their job, like `getSiteFragment`), so I'd alter this one to:

```
'/settings/disconnect-site/confirm/:site_id',
```

To reproduce an error with the `:site_id` not being at the end, while in the disconnect confirm page, switch your site. You'll be redirected to `settings/disconnect-site/confirm/:site` (notice where the site slug is).

File: client/my-sites/site-settings/disconnect-site/confirm.jsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+const ConfirmDisconnection = ( { siteId, siteSlug, translate } ) => {
+	if ( ! siteId ) {
+		return <Placeholder />;
+	}
+
+	return (
+		<Main className="disconnect-site__confirm">
+			<DocumentHead title={ translate( 'Site Settings' ) } />
+			<FormattedHeader
```
Comment (tyxla) [inline client/my-sites/site-settings/disconnect-site/confirm.jsx]: This is rather a design comment, but I was just thinking: don't we have just too many headings/subheadings here?



cc @keoshi

File: client/my-sites/site-settings/disconnect-site/confirm.jsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+				headerText={ translate( 'Confirm Disconnection' ) }
+				subHeaderText={ translate(
+					'Confirm that you want to disconnect your site from WordPress.com.'
+				) }
+			/>
+			<DisconnectJetpack
+				disconnectHref="/stats"
+				stayConnectedHref={ '/settings/manage-connection/' + siteSlug }
```
Comment (tyxla) [inline client/my-sites/site-settings/disconnect-site/confirm.jsx]: (nit) ABC police: would you also like to put props in alphabetical order? 😉

File: client/my-sites/site-settings/index.js (hunk 1 above)
Comment (ockham) [reply client/my-sites/site-settings/index.js]: Ah, good point. I modelled this after the wizard in https://github.com/Automattic/wp-calypso/blob/30d4eacd8a0b88cf2f8eaea766b85db418900506/client/extensions/wp-job-manager/index.js#L44, but yeah, the reason/confirm fragment should proably be next to `disconnect-site`.

File: client/my-sites/site-settings/disconnect-site/confirm.jsx (hunk 2 above)
Comment (keoshi) [reply client/my-sites/site-settings/disconnect-site/confirm.jsx]: @tyxla we totally do, this should be adjusted.

Comment (ockham) [reply client/my-sites/site-settings/disconnect-site/confirm.jsx]: I found this mockup over in the PR desc for #18039

screen shot 2017-09-13 at 20 30 59

But it diverges quite a bit from the dialog, especially regarding the layout of the "By disconnection..." part...

Comment (ockham) [reply client/my-sites/site-settings/disconnect-site/confirm.jsx]: Also, I guess we want full-screen (i.e. no sidebar)?

Comment (keoshi) [reply client/my-sites/site-settings/disconnect-site/confirm.jsx]: @ockham I would say let's leave the big paragraph as is (without the red icon) and just remove the heading. Also, yes to full-screen!

File: client/my-sites/site-settings/disconnect-site/style.scss [hunk 4]
```diff
@@ -1,34 +1,33 @@
+/** @format */
 .disconnect-site {
 	&.site-settings {
-	position: fixed;
-  top: 40%;
-  left: 50%;
-  transform: translate(-50%, -50%);
+		// Bit hacky -- compensate the padding added for the sidebar (which we aren't showing here).
+		margin-left: -340px;
```
Comment (sirreal) [inline client/my-sites/site-settings/disconnect-site/style.scss]: I've done a bit of Calypso archeology and I've got a "less-hacky" or at least "more-Calypso-ey" solution:

```patch
diff --git a/client/my-sites/site-settings/controller.js b/client/my-sites/site-settings/controller.js
index 7084a3000e..1376b14a33 100644
--- a/client/my-sites/site-settings/controller.js
+++ b/client/my-sites/site-settings/controller.js
@@ -26,6 +26,7 @@ import { getSelectedSite, getSelectedSiteId, getSelectedSiteSlug } from 'state/u
 import { isJetpackSite } from 'state/sites/selectors';
 import { canCurrentUser, isVipSite } from 'state/selectors';
 import { SITES_ONCE_CHANGED } from 'state/action-types';
+import { setSection } from 'state/ui/actions';
 
 function canDeleteSite( state, siteId ) {
 	const canManageOptions = canCurrentUser( state, siteId, 'manage_options' );
@@ -111,11 +112,13 @@ const controller = {
 
 	disconnectSite( context ) {
 		ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
+		context.store.dispatch( setSection( null, { hasSidebar: false } ) );
 		renderPage( context, <DisconnectSite /> );
 	},
 
 	disconnectSiteConfirm( context ) {
 		ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
+		context.store.dispatch( setSection( null, { hasSidebar: false } ) );
 		renderPage( context, <ConfirmDisconnection /> );
 	},
```

This sets the `has-no-sidebar` class on the reactroot, which takes care of the sidebar spacing. You can see it live by dispatching:

```js
dispatch( { type: 'SECTION_SET', hasSidebar: false });
```

Comment (ockham) [reply client/my-sites/site-settings/disconnect-site/style.scss]: Hah, I knew there had got to be a better way! Thanks a million Jon!

**Comment by sirreal** [comment]
If you visit with no site slug, you get a permanent placeholder. Probably not a major concern, but I don't know if `siteSelection` or a redirect should be included:

https://calypso.live/settings/disconnect-site/confirm?branch=add/disconnection-flow-confirm-page

placeholder

**Comment by ockham** [comment]
> If you visit with no site slug, you get a permanent placeholder. Probably not a major concern, but I don't know if `siteSelection` or a redirect should be included:

Good spot! This is what the `sites` middleware is for: https://github.com/Automattic/wp-calypso/pull/18796/commits/a431f6d5d3f2e120e633964b6ef33cca06d2c250

**Comment by ockham** [comment]
Added a prop to conditionally hide the title, as suggested here: https://github.com/Automattic/wp-calypso/pull/18796#pullrequestreview-69263509

Not sure it's the most elegant way of doing it, but since the title differs depending on whether `isBroken` is `true` or `false`, and due to the nesting of components inside a `Card`, and in case of the dialog, in a `Dialog` component (and related styling questions), it seemed like the easiest option to me.

File: client/my-sites/site-settings/index.js [hunk 5]
```diff
@@ -57,12 +57,21 @@ export default function() {
 	);
 
 	if ( config.isEnabled( 'manage/site-settings/disconnect-flow' ) ) {
+		page( '/settings/disconnect-site/:step(confirm)?', mySitesController.sites );
```
Comment (sirreal) [inline client/my-sites/site-settings/index.js]: Is this to cover the case that shows a placeholder forever? Is `:step(confirm)?` necessary, or could it just be `confirm?`.

Not a big deal, just curious 🙂

Comment (ockham) [reply client/my-sites/site-settings/index.js]: #18749 will add other possible `steps` :slightly_smiling_face:

**Pull request merged**